### ISSUE TITLE
future painting

### ISSUE THREAD

**Issue by macd2** [Fsa55/easybt#1]
i like your approach but it looks like your code is suffering from future painting, meaning the backtester can see in the future what in a real word is ofcause not possible

**Comment by Fsa55** [comment]
Yeah, it's more like "first try" project and was intended as something for fast rough estimation with subsequent analysis in tick by tick  backtester, which was done by IT department in company I worked for (as trader). The idea in readme example is that we enter by close price during the candle (period) where signal occured and, yes, entering position like this is a bit of cheating. But it is possible to override strategy method and use next open price to define the price of the position, create custom delay or anything like this. This is my first python project, so I guess more issues will come up, but speaking of default position enter: I will add option to choose wether to enter on close or next open soon.

**Comment by macd2** [comment]
PS: the future painting i was mentioning earlier happens here: 
 def run(self):
        """
        Start Backtest
        """
        **self.preprocess()**
        ....

You first calculate the signals for the entire df and than go over it row by row. This doesn't make any sense. You should redo the signal calculation on each time step....
